File: swift/Tests/allmarkTests/RenderToConsoleTests.swift
import Testing
@testable import allmark

@Test func renderConsoleParagraph() async {
	let input = "Hello, world!"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output == "Hello, world!")
	}
}

@Test func renderConsoleHeading() async {
	let input = "# Heading 1\n## Heading 2"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("# Heading 1"))
		#expect(output.contains("## Heading 2"))
	}
}

@Test func renderConsoleBulletedList() async {
	let input = "- Item 1\n- Item 2"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("• Item 1"))
		#expect(output.contains("• Item 2"))
	}
}

@Test func renderConsoleOrderedList() async {
	let input = "1. First\n2. Second"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("1. First"))
		#expect(output.contains("2. Second"))
	}
}

@Test func renderConsoleCodeBlock() async {
	let input = "```\ncode\n```"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("┌─"))
		#expect(output.contains("│"))
		#expect(output.contains("└─"))
	}
}

@Test func renderConsoleInlineCode() async {
	let input = "`code`"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("`code`"))
	}
}

@Test func renderConsoleBlockQuote() async {
	let input = "> Quote text"
 
	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		var stripped = ""
		var i = output.startIndex
		while i < output.endIndex {
			let char = output[i]
			if char == "\u{001B}" {
				let j = output.index(i, offsetBy: 1)
				if j < output.endIndex && output[j] == "[" {
					var k = output.index(j, offsetBy: 1)
					while k < output.endIndex {
						let nextChar = output[k]
						if nextChar == "m" {
							i = output.index(k, offsetBy: 1)
							break
						}
						if nextChar != ";" && !("0"..."9").contains(nextChar) {
							i = k
							break
						}
						k = output.index(k, offsetBy: 1)
					}
					continue
				}
			}
			stripped.append(char)
			i = output.index(i, offsetBy: 1)
		}
		#expect(stripped.contains("┃ Quote text"))
	}
}

@Test func renderConsoleThematicBreak() async {
	let input = "---"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("─"))
	}
}

@Test func renderConsoleTaskList() async {
	let input = "- [x] Done\n- [ ] Todo"
 
	await MainActor.run {
		let root = parse(src: input, rules: gfmRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: gfmRuleSet)
		#expect(output.contains("[✓]"))
		#expect(output.contains("[ ]"))
	}
}

@Test func renderConsoleTable() async {
	let input = "| A | B |\n|---|---|\n| 1 | 2 |"

	await MainActor.run {
		let root = parse(src: input, rules: gfmRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: gfmRuleSet)
		#expect(output.contains("┌"))
		#expect(output.contains("┬"))
		#expect(output.contains("┐"))
		#expect(output.contains("┼"))
		#expect(output.contains("│"))
		#expect(output.contains("├"))
		#expect(output.contains("┤"))
		#expect(output.contains("└"))
		#expect(output.contains("┴"))
		#expect(output.contains("┘"))
	}
}

@Test func renderConsoleStrongText() async {
	let input = "**bold**"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("bold"))
	}
}

@Test func renderConsoleEmphasisText() async {
	let input = "*italic*"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("italic"))
	}
}

@Test func renderConsoleLink() async {
	let input = "[text](url)"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("text"))
		#expect(output.contains("url"))
	}
}

@Test func renderConsoleImage() async {
	let input = "![alt](url)"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("Image"))
	}
}

@Test func renderConsoleStrikethrough() async {
	let input = "~~deleted~~"

	await MainActor.run {
		let root = parse(src: input, rules: gfmRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: gfmRuleSet)
		#expect(output.contains("~~deleted~~"))
	}
}

@Test func renderConsoleAlert() async {
	let input = "> [!NOTE]\n> Note content"

	await MainActor.run {
		let root = parse(src: input, rules: gfmRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: gfmRuleSet)
		#expect(output.contains("📝"))
		#expect(output.contains("Note:"))
	}
}

@Test func renderConsoleNestedList() async {
	let input = "- Level 1\n  - Level 2\n    - Level 3"

	await MainActor.run {
		let root = parse(src: input, rules: coreRuleSet, debug: false)
		let output = renderToConsole(doc: root, rules: coreRuleSet)
		#expect(output.contains("• Level 1"))
		#expect(output.contains("◦ Level 2"))
	}
}

File: dotnet/Allmark.Tests/RenderToConsoleTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Allmark.Rulesets;

namespace Allmark.Tests;

[TestClass]
public class RenderToConsoleTests
{
    [TestMethod]
    public void RenderConsoleParagraph()
    {
        var input = "Hello, world!";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("Hello, world!"));
    }

    [TestMethod]
    public void RenderConsoleHeading()
    {
        var input = "# Heading 1\n## Heading 2";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("# Heading 1"));
        Assert.IsTrue(output.Contains("## Heading 2"));
    }

    [TestMethod]
    public void RenderConsoleBulletedList()
    {
        var input = "- Item 1\n- Item 2";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("• Item 1"));
        Assert.IsTrue(output.Contains("• Item 2"));
    }

    [TestMethod]
    public void RenderConsoleOrderedList()
    {
        var input = "1. First\n2. Second";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("1. First"));
        Assert.IsTrue(output.Contains("2. Second"));
    }

    [TestMethod]
    public void RenderConsoleCodeBlock()
    {
        var input = "```\ncode\n```";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("┌─"));
        Assert.IsTrue(output.Contains("│"));
        Assert.IsTrue(output.Contains("└─"));
    }

    [TestMethod]
    public void RenderConsoleThematicBreak()
    {
        var input = "---";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("─"));
    }

    [TestMethod]
    public void RenderConsoleTaskList()
    {
        var input = "- [x] Done\n- [ ] Todo";
        var root = Parser.Execute(input, Gfm.RuleSet, false);
        var output = RenderToConsole.Execute(root, Gfm.RuleSet);
        Assert.IsTrue(output.Contains("[✓]"));
        Assert.IsTrue(output.Contains("[ ]"));
    }

    [TestMethod]
    public void RenderConsoleStrongText()
    {
        var input = "**bold**";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("bold"));
    }

    [TestMethod]
    public void RenderConsoleEmphasisText()
    {
        var input = "*italic*";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("italic"));
    }

    [TestMethod]
    public void RenderConsoleLink()
    {
        var input = "[text](url)";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("text"));
        Assert.IsTrue(output.Contains("url"));
    }

    [TestMethod]
    public void RenderConsoleImage()
    {
        var input = "![alt](url)";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("Image"));
    }

    [TestMethod]
    public void RenderConsoleStrikethrough()
    {
        var input = "~~deleted~~";
        var root = Parser.Execute(input, Gfm.RuleSet, false);
        var output = RenderToConsole.Execute(root, Gfm.RuleSet);
        Assert.IsTrue(output.Contains("\x1b[9mdeleted\x1b[29m"));
    }

    [TestMethod]
    public void RenderConsoleAlert()
    {
        var input = "> [!NOTE]\n> Note content";
        var root = Parser.Execute(input, Gfm.RuleSet, false);
        var output = RenderToConsole.Execute(root, Gfm.RuleSet);
        Assert.IsTrue(output.Contains("📝"));
        Assert.IsTrue(output.Contains("Note:"));
    }

    [TestMethod]
    public void RenderConsoleNestedList()
    {
        var input = "- Level 1\n  - Level 2\n    - Level 3";
        var root = Parser.Execute(input, Core.RuleSet, false);
        var output = RenderToConsole.Execute(root, Core.RuleSet);
        Assert.IsTrue(output.Contains("• Level 1"));
        Assert.IsTrue(output.Contains("◦ Level 2"));
    }
}
